File: Dynamic_Programming/1932.cpp
#include <iostream>

using namespace std;

int max(int x,int y);

int main(){
    int DP[500][500]={0,};
    int n=0;
    int M=0;
    cin>>n;//삼각형의 높이

    for(int k=0;k<n;k++) {
        for (int i = 0; i <= k; i++){
            cin>>DP[k][i];
        }
    }

    for(int k=1;k<n;k++) {
        for (int i = 0; i <=k; i++){
            if(i==0){
                DP[k][i]+=DP[k-1][i];
            }
            else if(i==k){
                DP[k][i]+=DP[k-1][i-1];
            }
            else
                DP[k][i]+=max(DP[k-1][i],DP[k-1][i-1]);

            if(M<DP[k][i])
                M=DP[k][i];
        }
    }

    cout<< M;

    return 0;
}

int max(int x,int y){
    if(x>y||x==y)
        return x;
    else
        return y;

}

File: Dynamic_Programming/1932.py
import sys

def solution(n):
    result=0
    triangle=list()
    triangle=[[0 for col in range(n+1)] for row in range(n+1)]
    DP=[[0 for col in range(n+1)] for row in range(n+1)]

    for i in range(n):
        triangle[i]=list(map(int,sys.stdin.readline().split()))

    DP[0][0]=triangle[0][0]
    DP[1][0]=DP[0][0]+triangle[1][0]
    DP[1][1]=DP[0][0]+triangle[1][1]

    for layer in range(2,n):
        for idx in range(layer+1):
            if idx == 0:
                DP[layer][idx]=triangle[layer][idx]+DP[layer-1][0]
            elif idx == layer:
                DP[layer][idx]=triangle[layer][idx]+DP[layer-1][idx-1]
            else:
                DP[layer][idx]=triangle[layer][idx]+max(DP[layer-1][idx],DP[layer-1][idx-1])


    for i in range(n):
        result=max(result,DP[n-1][i])

    return result



# Press the green button in the gutter to run the script.
if __name__ == '__main__':
    n=int(input())
    print(solution(n))
# See PyCharm help at https://www.jetbrains.com/help/pycharm/
